vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListController.kt

```diff
@@ -0,0 +1,236 @@
+
+package fr.gouv.tchap.features.home.contact.list
+
+import com.airbnb.epoxy.EpoxyController
+import com.airbnb.mvrx.Fail
+import com.airbnb.mvrx.Loading
+import com.airbnb.mvrx.Success
+import com.airbnb.mvrx.Uninitialized
+import im.vector.app.R
+import im.vector.app.core.contacts.MappedContact
+import im.vector.app.core.epoxy.errorWithRetryItem
+import im.vector.app.core.epoxy.loadingItem
+import im.vector.app.core.epoxy.noResultItem
+import im.vector.app.core.error.ErrorFormatter
+import im.vector.app.core.resources.StringProvider
+import im.vector.app.features.contactsbook.contactDetailItem
+import im.vector.app.features.contactsbook.contactItem
+import im.vector.app.features.home.AvatarRenderer
+import im.vector.app.features.userdirectory.userDirectoryUserItem
+import im.vector.app.features.userdirectory.userListHeaderItem
+import org.matrix.android.sdk.api.session.Session
+import org.matrix.android.sdk.api.session.user.model.User
+import org.matrix.android.sdk.api.util.toMatrixItem
+import javax.inject.Inject
+
+class TchapContactListController @Inject constructor(private val session: Session,
+                                                     private val avatarRenderer: AvatarRenderer,
+                                                     private val stringProvider: StringProvider,
+                                                     private val errorFormatter: ErrorFormatter) : EpoxyController() {
+
+    private var state: TchapContactListViewState? = null
+
+    var callback: Callback? = null
+
+    fun setData(state: TchapContactListViewState) {
+        this.state = state
+        requestModelBuild()
+    }
+
+    override fun buildModels() {
+        val currentState = state ?: return
+
+        when (currentState.roomSummaries) {
+            is Uninitialized -> renderEmptyState()
+            is Loading -> renderLoading()
+            is Fail -> renderFailure(currentState.roomSummaries.error)
+            is Success -> buildlocalContacts(currentState)
+        }
+
+        when (val asyncUsers = currentState.directoryUsers) {
+            is Uninitialized -> {
+            }
+            is Loading -> renderLoading()
+            is Fail -> renderFailure(asyncUsers.error)
+            is Success -> buildDirectoryUsers(
+                    asyncUsers(),
+                    currentState.searchTerm,
+                    emptyList()
+            )
+        }
+    }
+
+    private fun buildlocalContacts(currentState: TchapContactListViewState) {
+        val userList = currentState.filteredRoomSummaries.mapNotNull { item ->
+            item.directUserId?.let {
+                User(it, item.displayName, item.avatarUrl)
+            }
+        }
+        val tchapContactList: MutableList<Any> = userList.toMutableList()
+
+        tchapContactList.addAll(currentState.filteredLocalContacts.filter { mappedContact ->
+            mappedContact.emails.any { it.matrixId != null && !userList.any { user -> user.userId == it.matrixId } }
```

Comment:
I think you can simplify like this:
mappedContact.emails
    .filterNotNull { it.matrixId }
    .any { userList.none { user -> user.userId == it.matrixId }

Reply:
Why not but in your case you iterate two times (filterNotNull + any) on the same elements.